[FLINK-11545] [container] Add job-id arg to StandaloneJobClusterEntryPoint

uce opened the pull request:
## What is the purpose of the change

- This PR adds the optional `--job-id` command line argument to the `StandaloneJobClusterEntryPoint`
  - If specified, users manually set the job ID of the job
  - If not specified, the job ID of the job is all zeros (current behavior)

## Brief change log

- Parse job ID option in `StandaloneJobClusterConfigurationParserFactory` 
- Provide job ID in `ClassPathJobGraphRetriever` constructor
- Catch `Throwable` when parsing command line args as a safety net and have a good error message in the logs

## Verifying this change

- This change added unit tests for added functionality
- If you want to manually verify this change, you can follow these steps:
  1. Build this branch
  2. Copy `examples/streaming/TopSpeedWindowing.jar` to `lib`
  3. Execute `bin/standalone-job.sh start --job-classname org.apache.flink.streaming.examples.windowing.TopSpeedWindowing` (expected: job ID is all zeros)
  4. Execute `bin/standalone-job.sh start --job-classname org.apache.flink.streaming.examples.windowing.TopSpeedWindowing --job-id fd72014d4c864993a2e5a9287b4a9c5d` (expected: job ID fd72014d4c864993a2e5a9287b4a9c5d)
  5. Execute `bin/standalone-job.sh start --job-classname org.apache.flink.streaming.examples.windowing.TopSpeedWindowing --job-id INVALID_JOB_ID` (expected: logs with error message)

## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): no
- The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes (it adds a new command line argument which is part of the extended API)
- The serializers: no
- The runtime per-record code paths (performance sensitive): no
- Anything that affects deployment or recovery: no
- The S3 file system connector: no

## Documentation

- Added note about new command line arg in `flink-container/kubernetes/README.md`

uce commented:
CC @yanghua @tillrohrmann

zentol commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java:
```diff
@@ -79,6 +92,20 @@ public StandaloneJobClusterConfiguration createResult(@Nonnull CommandLine comma
+		try {
+			return JobID.fromHexString(jobId);
+		} catch (Throwable t) {
+			throw new IllegalArgumentException("Failed to parse job ID '" + jobId + "'", t);
```
add a description of what format is expected

uce replied:
Good idea. Is the updated message OK?

zentol commented on flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java:
```diff
@@ -97,4 +103,27 @@ public void testSavepointRestoreSettingsParsing() throws FlinkParseException {
+	@Test
+	public void testSetJobIdManually() throws FlinkParseException {
+		final JobID jobId = new JobID();
+		final String[] args = {"--configDir", "/foo/bar", "--job-classname", "foobar", "--job-id", jobId.toString()};
```
please also test the short option version `-jid`

uce replied:
What are we testing with this? That the name `jid` does not change in the future? I don't want to test the library that we are using.

zentol replied:
if `-jid` is an accepted command-line argument then I would consider it as part of the API, in wbhicih case it should be covered by a test.

Is this just an artifact from the library we're using?

uce commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java:
```diff
@@ -37,10 +40,14 @@
+	@Nonnull
+	private final JobID jobId;
+
+	StandaloneJobClusterConfiguration(@Nonnull String configDir, @Nonnull Properties dynamicProperties, @Nonnull String[] args, @Nullable String hostname, int restPort, @Nonnull String jobClassName, @Nonnull SavepointRestoreSettings savepointRestoreSettings, @Nonnull JobID jobId) {
```
Sure/ What style do we follow here? One argument per line if we break the line or just multiple lines with 120 characters.

uce replied:
Picked one line per argument.

uce commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java:
```diff
@@ -79,6 +92,21 @@ public StandaloneJobClusterConfiguration createResult(@Nonnull CommandLine comma
+		try {
+			return JobID.fromHexString(jobId);
+		} catch (Throwable t) {
+			throw new IllegalArgumentException("Failed to parse job ID '" + jobId + "'. The expected format is " +
```
I think the best option would be to update the `JobID.fromHexString` message.

uce commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java:
```diff
@@ -33,38 +33,44 @@
  */
 public final class StandaloneJobClusterEntryPoint extends JobClusterEntrypoint {
 
-	private final String[] programArguments;
```
Agree. But I think we typically align order of member variables and constructor arguments. Do you still want me to have the old order?

zentol replied:
you can move it into another `[hotfix]` commit.

uce commented on flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java:
```diff
@@ -78,8 +84,8 @@ public static void main(String[] args) {
 			clusterConfiguration = commandLineParser.parse(args);
-		} catch (FlinkParseException e) {
-			LOG.error("Could not parse command line arguments {}.", args, e);
+		} catch (Throwable t) {
```
Should not make any difference here, but works for me. The important bit is that we want to have the log code path in order to have the error message in the log.

uce commented:
@zentol Thanks for the review. 👍 I addressed most of the comments. I have two open questions that I will address after your answers.

uce commented:
@zentol I added a `[hotfix]` commit for the reordering (before all changes of this PR) and added a short options test.

uce commented:
I rebased this PR on master and cleaned up the review commit history. Thanks for your reviews Chesnay. I'll merge this after Travis gives a green light.
